url6a9f553762_Bresiu_socket_io_tgz-pJ8-MainJ8/src/com/github/nkzawa/engineio/parser/Parser.java

```diff
@@ -0,0 +1,305 @@
+        int type;
+        try {
+            type = Character.getNumericValue(data.charAt(0));
+        } catch (IndexOutOfBoundsException e) {
+            type = -1;
+        }
+        try {
+            data = UTF8.decode(data);
+        } catch (UTF8Exception e) {
+            return err;
+        }
+
+        if (type < 0 || type >= packetslist.size()) {
+            return err;
+        }
+
+        if (data.length() > 1) {
+            return new Packet<String>(packetslist.get(type), data.substring(1));
+        } else {
+            return new Packet<String>(packetslist.get(type));
+        }
+    }
+
+    public static Packet<byte[]> decodePacket(byte[] data) {
+        int type = data[0];
+        byte[] intArray = new byte[data.length - 1];
+        System.arraycopy(data, 1, intArray, 0, intArray.length);
+        return new Packet<byte[]>(packetslist.get(type), intArray);
+    }
+
+    public static void encodePayload(Packet[] packets, EncodeCallback<byte[]> callback) {
+        if (packets.length == 0) {
+            callback.call(new byte[0]);
+            return;
+        }
+
+        final ArrayList<byte[]> results = new ArrayList<byte[]>(packets.length);
+
+        for (Packet packet : packets) {
+            encodePacket(packet, new EncodeCallback() {
+                @Override
+                public void call(Object packet) {
+                    if (packet instanceof String) {
+                        String encodingLength = String.valueOf(((String) packet).length());
+                        byte[] sizeBuffer = new byte[encodingLength.length() + 2];
+
+                        sizeBuffer[0] = (byte)0; // is a string
+                        for (int i = 0; i < encodingLength.length(); i ++) {
+                            sizeBuffer[i + 1] = (byte)Character.getNumericValue(encodingLength.charAt(i));
+                        }
+                        sizeBuffer[sizeBuffer.length - 1] = (byte)255;
+                        results.add(Buffer.concat(new byte[][] {sizeBuffer, stringToByteArray((String)packet)}));
+                        return;
+                    }
+
+                    String encodingLength = String.valueOf(((byte[])packet).length);
+                    byte[] sizeBuffer = new byte[encodingLength.length() + 2];
+                    sizeBuffer[0] = (byte)1; // is binary
+                    for (int i = 0; i < encodingLength.length(); i ++) {
+                        sizeBuffer[i + 1] = (byte)Character.getNumericValue(encodingLength.charAt(i));
+                    }
+                    sizeBuffer[sizeBuffer.length - 1] = (byte)255;
+                    results.add(Buffer.concat(new byte[][] {sizeBuffer, (byte[])packet}));
+                }
+            });
+        }
+
+        callback.call(Buffer.concat(results.toArray(new byte[results.size()][])));
+    }
+
+    public static void decodePayload(String data, DecodePayloadCallback<String> callback) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.